### ISSUE TITLE
maketx -u should ignore paths

### ISSUE THREAD

**Issue by ThiagoIze** [OpenImageIO/oiio#1472]
maketx -u  will store the command used to run it and then use that in the future to see if the texture should be updated.  This is overly conservative because the paths can benignly differ between uses.  For instance, the following might all be the same but maketx -u will consider them all different:

`/Users/JohnDoe/oiio/maketx -u /full/path/to/file/on/server/foo.jpg`
`/Users/OtherPerson/oiio/maketx -u /full/path/to/file/on/server/foo.jpg`
`/Users/OtherPerson/oiio/maketx -u k:\full\path\to\file\on\server\foo.jpg`
`maketx -u foo.jpg`

The different paths based on OS is especially problematic since I can't see of a reasonable way to avoid that for people that have a heterogeneous mix of machines.

My suggestion for how to solve this is that all the paths could be stripped of the leading parts of the path so that only the final filename is used (what the `basename` tool does).  In the above example, all of these commands would simplify to:  `maketx -u foo.jpg`. 

In fact, it might even be possible to completely get rid of the entire path. I don't think we actually need the filename for the source and destination and the name of the executable is also not needed since we already have the version number encoded in the "Software" string we're examining.  I haven't gone through all the other paths in maketx, but it's possible the others could also be entirely ignored?

This isn't 100% foolproof, for instance, on case insensitive filesystems foo.jpg and Foo.JPG are the same, but I hope that simply removing parts or entirety of paths will be good enough.

**Comment by lgritz** [comment]
I think I agree about the path to maketx, though I wonder how often it would really come up. Because we encode the version separately, I think you're right that the full path to the maketx binary probably doesn't give us any additional safety. I suppose we can use just the basename? Or make an option to do so?

I'm somewhat less convinced about the path to the input filename. That seems like it could have problems with false negatives. The only problem with a false positive that you describe is that it might spend an extra couple of seconds re-building the texture. But a false negative could be bad. As an example, consider a first maketx of /textures/v1/x.exr and then a later /textures/v2/x.exr, which are intentionally different versions. Wouldn't want to skip the maketx in that case, right?

Also, to be honest, I'm loathe to have to put a bunch of string parsing and rewriting into this comparison, and I worry about making mistakes that botch it, especially for Windows filename corner cases that I might not think of.

Another possibility, and a way of sweeping the path munging under the rug, is not using the "Software" metadata at all for maketx update mode, and *strictly* relying on the file dates (or having an option to do so). That would put more of a burden on users to be very careful about how they use the `-u` option. Maybe if we want that, a second option is better for this?

**Comment by ThiagoIze** [comment]
The path to maketx issue was happening often for many of our customers, especially for those with mixed windows/linux usage. Better pipelines could avoid it, but that's often out of our control. basename would work. In our branch I think we ignore the whole thing (path and maketx), but I don't think that was necessary and I do realize that carries extra risks.

About the v1/x.exr vs v2/x.exr (with I assume a -o pointing to the same place), they are very likely to have different timestamps, so the update would still go through. If the timestamps are the same, then that's weird, just like it'd be weird to overwrite v1/x.exr with a new data but keep the timestamp the same.

We've had this in our branch for two years now, doing regular std::string stuff, and I don't believe we've had any complaints from our windows users, but I also would not be shocked to learn this fails in Windows for certain special locales. I'm happy to share with you our code, but I wrote it in order to keep the changes minimal (fewer merge conflicts) rather than doing more sweeping changes that result in cleaner code.

Just looking at the timestamps I think is too risky since changes to params, say a change to the filter being used, would not trigger an update. Asking the user have to keep track of these changes is rather challenging and at that point -u doesn't really provide much.

If it matters, since we already have this fixed in our branch, getting this also fixed upstream is low priority for us.

**Comment by lgritz** [comment]
v1 vs v2: Dunno about how copying works on Windows, I'm just thinking that `cp -r /oldshots /newshots` can end with a whole slew of files having the exact same modification time. It's oddness like that which made me add the extra test of making sure the same command line was used. (Solution for the Unix-savvy: `cp -rp`)

In your branch, did you modify maketx to strip the paths when first putting it into the metadata (for example, by putting the basename smarts in `command_line_string()`), or are the modifications to the way the strings are compared for the update mode logic in `make_texture_impl()`? The former seems a lot simpler to implement, but truth be told, at my studio I frequently use the fact that the full pathnames are preserved in the metadata to diagnose subtle mistakes such as people accidentally invoking the wrong copy of the software (like from a custom tree where they were experimenting with something and forgot to switch back to an official release).

If you have a fix you're willing to contribute back, I would be happy to receive it. I'm just a little uncertain of my ability to do the filename canonicalization myself without botching it for some odd Windows corner case, given that my ability to test thoroughly on Windows is extremely limited. Also, I don't want you be burdened any more than necessary with having to maintain patches on your side that you have to re-apply every time you upgrade. (Especially in the new "clang-format" days, when merging patches written against 1.x to a new 2.0 will be a more difficult merge than usual.)

Since nobody else has ever mentioned this issue, what I'm inclined to do is close this ticket, knowing that if you want to turn it into a PR that changes the behavior in a way that you think is better, I'll happily merge it upstream if it seems like it's reasonably safe.

**Comment by ThiagoIze** [comment]
If copying files gives the original timestamp, I don't think there's any problem -- both versions are identical copies of each other after all. It's not until you modify one of them that you'll want maketx to differentiate and at that point the modification should have updated the timestamp.

The change is in make_texture_impl so there's no change to what's written in the "Software" field.

Making a proper pull request that's guaranteed to compile and work I can't do at the moment (too many related and unrelated changes in our branch), but I can post the initial patch here that very likely will not directly apply but should be easy to at least see how it works if someone feels inspired to properly implement it:  [path_strip.diff.txt](https://github.com/OpenImageIO/oiio/files/2688532/path_strip.diff.txt). Feel free to close this ticket since I'm not likely to create a PR any time soon and at least on our end everything seems to be working for us.

**Comment by lgritz** [comment]
Thanks for the patch, but after looking over, I took a stab at re-implementing in an even simpler way. Curious about your opinion:  #2109
